ppoprf/src/ppoprf.rs

```diff
@@ -87,6 +149,152 @@ impl ProofDLEQ {
+    &self,
+    public_value: &RistrettoPoint,
+    p: &[RistrettoPoint], //P
+    q: &[RistrettoPoint], //Q
+  ) -> bool {
+    //(M, Z) = ComputeComposites(B, C, D)
+    let (m, z) = ProofDLEQ::compute_composites(None, public_value, p, q);
+
+    //t2 = ((s * A) + (c * B))
+    let t2 = (self.s * RISTRETTO_BASEPOINT_POINT) + (self.c * public_value);
+    //t3 = ((s * M) + (c * Z))
+    let t3 = (self.s * m) + (self.c * z);
+
+    //Bm = G.SerializeElement(B)
+    let bm_string = ProofDLEQ::serialize_element(*public_value);
+
+    /*a0 = G.SerializeElement(M)
+    a1 = G.SerializeElement(Z)
+    a2 = G.SerializeElement(t2)
+    a3 = G.SerializeElement(t3) */
+    let a0_string = ProofDLEQ::serialize_element(m);
+    let a1_string = ProofDLEQ::serialize_element(z);
+    let a2_string = ProofDLEQ::serialize_element(t2);
+    let a3_string = ProofDLEQ::serialize_element(t3);
+
+    /*challengeTranscript =
+    I2OSP(len(Bm), 2) || Bm ||
+    I2OSP(len(a0), 2) || a0 ||
+    I2OSP(len(a1), 2) || a1 ||
+    I2OSP(len(a2), 2) || a2 ||
+    I2OSP(len(a3), 2) || a3 ||
+    "Challenge"*/
+    let challenge_transcript = format!(
+      "{}{}{}{}{}{}{}{}{}{}",
+      bm_string.len(),
+      bm_string,
+      a0_string.len(),
+      a0_string,
+      a1_string.len(),
+      a1_string,
+      a2_string.len(),
+      a2_string,
+      a3_string.len(),
+      a3_string
+    );
+
+    //c = G.HashToScalar(challengeTranscript)
+    let mut out = [0u8; 64];
+    strobe_hash(&challenge_transcript.as_bytes(), "Challenge", &mut out);
+    let expected_c = RistrettoScalar::from_bytes_mod_order_wide(&out);
+
+    //verified = (expectedC == c)
+    return expected_c == self.c;
+  }
+
+  fn compute_composites(
+    key: Option<RistrettoScalar>,
+    b: &RistrettoPoint,
+    c: &[RistrettoPoint],
+    d: &[RistrettoPoint],
+  ) -> (RistrettoPoint, RistrettoPoint) {
+    if c.len() != d.len() {
+      panic!("C and D have a different number of elements!");
+    }
+
+    //Bm = G.SerializeElement(B)
+    let bm_string = ProofDLEQ::serialize_element(*b);
+
+    // We use the Partially-punctureable Oblivious Pseudo-Random Function
+    let context_string = format!(
+      "{}{}{}{}",
+      "PPOPRFv1-",
+      0x03.to_string(),
+      "-",
+      "ristretto255-strobe"
+    );
+    //seedDST = "Seed-" || contextString
+    let seed_dst = format!("{}{}", "Seed-", context_string);
+
+    // seedTranscript = I2OSP(len(Bm), 2) || Bm || I2OSP(len(seedDST), 2) || seedDST
+    let seed_transcript = format!(
+      "{}{}{}{}",
+      bm_string.len(),
+      bm_string,
+      seed_dst.len(),
+      seed_dst
+    );
+
+    let mut seed = [0u8; 64];
+    strobe_hash(&seed_transcript.as_bytes(), "Seed", &mut seed);
+
+    //M = G.Identity()
+    let mut m = RISTRETTO_BASEPOINT_POINT;
+    //Z = G.Identity()
+    let mut z = RISTRETTO_BASEPOINT_POINT;
+
+    let seed_string = match str::from_utf8(&seed) {
+      Ok(v) => v,
+      Err(e) => panic!("Invalid UTF-8 sequence: {}", e),
+    };
```

Comment:
I don't think this is what you want. There's no reason for the hash output to be valid utf8, so the panic can happen in the course of normal operation.

Instead I think you need to keep all of this as bytes. Allocate a `Vec<u8>` and use `.extend()` or similar to concatenate the various fields. The numbers can be serialized with something like `i.to_string().as_bytes()` but to follow the spec you should actually implement an `i2osp()` function to enforce the two-digit field size.

Then you can remove `serialize_element` entirely and just extend with `c[i].compress().as_bytes()` &c.

Reply:
FWIW there's another style here, where you use the `write!` macro to append things to a `Vec<u8>`. But that interface is fallible (since you might be writing to a file) so you need error handling.

For something this compact, I think working directly with a byte collection from std is the easiest to read.